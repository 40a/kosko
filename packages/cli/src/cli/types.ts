import { Logger } from "./logger";

export interface WithAlias {
  alias?: string | string[];
}

export interface BaseOption<T = any> {
  description?: string;
  default?: T | (() => T | undefined);
  required?: boolean;
  options?: T[];
}

export function getAliases({ alias }: WithAlias): string[] {
  if (Array.isArray(alias)) return alias;
  return alias ? [alias] : [];
}

export interface Arg<K = string, V = any> extends BaseOption<V> {
  name: K;
}

export enum OptionType {
  String = "string",
  Boolean = "boolean",
  Count = "count",
  Number = "number",
  Array = "array"
}

export interface Option<T = any> extends WithAlias, BaseOption<T> {
  type?: OptionType;
  group?: string;
}

export interface CommandMeta<Options = any> {
  usage?: string;
  description?: string;
  examples?: string;
  options?: { [key in keyof Options]: Option };
  commands?: { [key: string]: Command };
  args?: Arg[];
}

export interface Command<Options = any> extends CommandMeta<Options> {
  exec(ctx: Context, argv: string[]): void;
}

export interface Context {
  logger: Logger;
}
