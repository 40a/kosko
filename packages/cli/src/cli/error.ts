import chalk from "chalk";
import cleanStack from "clean-stack";
import exit from "exit";

export class CLIError extends Error {
  public readonly name = "CLIError";
  public readonly output?: string;
  public readonly code?: number;

  constructor(
    msg: string,
    { output, code }: { output?: string; code?: number } = {}
  ) {
    super(msg);
    this.output = output;
    this.code = code;
  }
}

export function formatError(err: Error) {
  if (!err.stack) return err.message;

  const stack = cleanStack(err.stack, { pretty: true });
  // Regular expression is from: https://github.com/sindresorhus/extract-stack
  const pos = stack.search(/(?:\n {4}at .*)+/);
  if (!~pos) return stack;

  return stack.substring(0, pos) + chalk.gray(stack.substring(pos));
}

export function handleError(err: any) {
  let code = 1;
  let msg = err;

  if (err instanceof CLIError) {
    msg = err.output || formatError(err);
    if (err.code != null) code = err.code;
  } else if (err instanceof Error) {
    msg = formatError(err);
  }

  // tslint:disable-next-line:no-console
  console.error(msg);
  exit(code);
}
