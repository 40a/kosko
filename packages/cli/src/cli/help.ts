import groupBy from "lodash.groupby";
import { Writable } from "stream";
import { Option, OptionType, CommandMeta } from "./types";

interface Entry {
  key: string;
  description: string;
}

interface OptionWithName extends Option {
  name: string;
}

function getOptionKey(opt: OptionWithName) {
  let key = [opt.name]
    .concat(opt.alias || [])
    .map(x => (x.length > 1 ? `--${x}` : `-${x}`))
    .sort((a, b) => a.length - b.length)
    .join(", ");

  switch (opt.type) {
    case OptionType.Boolean:
    case OptionType.Count:
      break;

    default:
      if (opt.type) key += ` ${opt.type}`;
  }

  return key;
}

class Help {
  private readonly cmd: CommandMeta;
  private readonly writer: Writable;

  private started = false;

  constructor(cmd: CommandMeta, writer: Writable) {
    this.cmd = cmd;
    this.writer = writer;
  }

  public async print() {
    await this.printUsage();
    await this.printDescription();
    await this.printArgs();
    await this.printCommands();
    await this.printOptions();
  }

  private write(chunk: any) {
    return new Promise((resolve, reject) => {
      this.writer.write(chunk, err => {
        if (err) return reject(err);
        resolve();
      });
    });
  }

  private async printUsage() {
    const { usage } = this.cmd;
    if (!usage) return;

    await this.printHeader("Usage: " + usage);
  }

  private async printDescription() {
    const { description } = this.cmd;
    if (!description) return;

    await this.printHeader(description);
  }

  private async printArgs() {
    const { args = [] } = this.cmd;

    await this.printEntries(
      "Arguments:",
      args.map(arg => ({ key: arg.name, description: arg.description || "" }))
    );
  }

  private async printCommands() {
    const { commands = {} } = this.cmd;
    const entries: Entry[] = [];

    for (const key of Object.keys(commands)) {
      entries.push({ key, description: commands[key].description || "" });
    }

    await this.printEntries("Commands:", entries);
  }

  private async printOptions() {
    const { options = {} } = this.cmd;
    const opts: OptionWithName[] = [];

    for (const key of Object.keys(options)) {
      opts.push({ ...options[key], name: key });
    }

    const groups = groupBy(opts, opt => opt.group || "Options");

    for (const key of Object.keys(groups)) {
      await this.printOptionGroup(key, groups[key]);
    }
  }

  private async printOptionGroup(group: string, options: OptionWithName[]) {
    await this.printEntries(
      `${group}:`,
      options.map(opt => ({
        key: getOptionKey(opt),
        description: opt.description || ""
      }))
    );
  }

  private async printHeader(header?: string) {
    if (this.started) await this.write("\n");
    await this.write(header + "\n");
    this.started = true;
  }

  private async printEntries(header: string, entries: Entry[]) {
    if (!entries.length) return;

    await this.printHeader(header);

    const keySize = Math.max(...entries.map(e => e.key.length));

    entries.sort((a, b) => {
      if (a.key > b.key) return 1;
      if (a.key < b.key) return -1;
      return 0;
    });

    for (const entry of entries) {
      const pad = Array(keySize - entry.key.length)
        .fill(" ")
        .join("");

      await this.write(`  ${entry.key}${pad}  ${entry.description}\n`);
    }
  }
}

export async function help<T = any>(
  cmd: CommandMeta<T>,
  writer: Writable = process.stdout
) {
  const helper = new Help(cmd, writer);
  return helper.print();
}
