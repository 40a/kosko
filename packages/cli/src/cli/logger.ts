import { Writable } from "stream";
import { format } from "util";

export class Logger {
  private readonly writer: Writable;

  constructor(writer: Writable) {
    this.writer = writer;
  }

  public log(msg: any, ...params: any[]) {
    this.writer.write(format(msg, ...params) + "\n");
  }
}
