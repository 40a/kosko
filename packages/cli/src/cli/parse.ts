import mapValues from "lodash.mapvalues";
import parser, { DetailedArguments } from "yargs-parser";
import {
  Arg,
  BaseOption,
  Command,
  getAliases,
  Option,
  OptionType
} from "./types";

export class ParseError extends Error {
  constructor(details: ParseErrorDetail[]) {
    const msg = details
      .map(({ arg, option, message }) => {
        if (arg) return `Argument "${arg}": ${message}`;
        return `Option "--${option}": ${message}`;
      })
      .join("\n");

    super(msg);
  }
}

export interface ParseErrorDetail {
  arg?: string;
  option?: string;
  message: string;
}

export interface ParseResult<Options, Args> {
  options: Options;
  args: Args;
  detail: DetailedArguments;
  errors: ParseErrorDetail[];
}

interface OptionMap {
  [key: string]: Option;
}

function filterByType(options: OptionMap, type: OptionType): string[] {
  return Object.keys(options).filter(key => options[key].type === type);
}

function getParserOptions(options: { [key: string]: Option }): parser.Options {
  return {
    alias: mapValues(options, getAliases),
    array: filterByType(options, OptionType.Array),
    boolean: filterByType(options, OptionType.Boolean),
    count: filterByType(options, OptionType.Count),
    number: filterByType(options, OptionType.Number),
    string: filterByType(options, OptionType.String)
  };
}

interface ValidateBaseOptionsResult {
  value?: any;
  error?: string;
}

function validateBaseOption(
  option: BaseOption,
  value: any
): ValidateBaseOptionsResult {
  if (value == null) {
    if (typeof option.default === "function") {
      return { value: option.default() };
    }

    if (option.default != null) {
      return { value: option.default };
    }

    if (option.required) {
      return { error: "Required" };
    }
  }

  if (Array.isArray(option.options)) {
    if (!option.options.includes(value)) {
      return { error: `Must be one of [${option.options.join(", ")}]` };
    }
  }

  return { value };
}

interface ValidateOptionsResult {
  options: any;
  errors: ParseErrorDetail[];
}

function validateOptions(
  detail: DetailedArguments,
  options: OptionMap
): ValidateOptionsResult {
  const result: ValidateOptionsResult = {
    options: {},
    errors: []
  };

  Object.keys(options).forEach(key => {
    const { value, error } = validateBaseOption(options[key], detail.argv[key]);

    if (error) {
      result.errors.push({ option: key, message: error });
    } else {
      result.options[key] = value;
    }
  });

  return result;
}

interface ValidateArgsResult {
  args: any;
  errors: ParseErrorDetail[];
}

function validateArgs(
  detail: DetailedArguments,
  args: Arg[]
): ValidateArgsResult {
  const result: ValidateArgsResult = {
    args: {},
    errors: []
  };

  args.forEach((arg, i) => {
    const { value, error } = validateBaseOption(arg, detail.argv._[i]);

    if (error) {
      result.errors.push({ arg: arg.name, message: error });
    } else {
      result.args[arg.name] = value;
    }
  });

  return result;
}

export function parse<Options, Args>(
  argv: string[],
  cmd: Pick<Command<Options>, "options" | "args">,
  config: Partial<parser.Configuration> = {}
): ParseResult<Options, Args> {
  const detail = parser.detailed(argv, {
    configuration: config,
    ...getParserOptions(cmd.options || {})
  });

  const { options, errors: optErrors } = validateOptions(
    detail,
    cmd.options || {}
  );
  const { args, errors: argErrors } = validateArgs(detail, cmd.args || []);

  return {
    detail,
    options,
    args,
    errors: [...optErrors, ...argErrors]
  };
}
