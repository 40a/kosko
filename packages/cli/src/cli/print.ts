export function print(data: string | Buffer) {
  return new Promise((resolve, reject) => {
    process.stdout.write(data, (err: any) => {
      if (err) return reject(err);
      resolve();
    });
  });
}
