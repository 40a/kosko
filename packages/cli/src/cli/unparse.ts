import { DetailedArguments } from "yargs-parser";

function getFlag(key: string, value: any) {
  const prefix = key.length > 1 ? "--" : "-";
  return prefix + key;
}

function flatten<T>(arr: T[][]): T[] {
  return arr.reduce((acc, x) => acc.concat(x), []);
}

function unparseOption(key: string, value: any, prefix: string = ""): string[] {
  const fullKey = prefix + key;

  if (value === true) return [getFlag(fullKey, value)];
  if (value === false) return [`--no-${fullKey}`];
  if (typeof value !== "object") return [getFlag(fullKey, value), `${value}`];

  let output: string[][];

  if (Array.isArray(value)) {
    output = value.map(v => unparseOption(key, v, prefix));
  } else {
    output = Object.keys(value).map(k =>
      unparseOption(k, value[k], `${prefix}${key}.`)
    );
  }

  return flatten(output);
}

function unparseOptions(args: { [key: string]: any }): string[] {
  const output = Object.keys(args).map(key => unparseOption(key, args[key]));
  return flatten(output);
}

function filterAliases(
  detail: DetailedArguments,
  args: { [key: string]: any }
) {
  const output: { [key: string]: any } = {};
  let filtered: string[] = [];

  for (const key of Object.keys(args)) {
    if (detail.aliases[key]) {
      filtered = filtered.concat(detail.aliases[key]);
    }

    if (!filtered.includes(key)) {
      output[key] = args[key];
    }
  }

  return output;
}

export function unparse(detail: DetailedArguments): string[] {
  const { _: positionals, $0, "--": ends, ...args } = detail.argv;

  return [
    ...positionals,
    ...unparseOptions(filterAliases(detail, args)),
    ...(Array.isArray(ends) ? ["--", ...ends] : [])
  ];
}
