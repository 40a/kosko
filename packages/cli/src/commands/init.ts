import Debug from "debug";
import fs from "fs";
import makeDir from "make-dir";
import { join, resolve } from "path";
import writePkg from "write-pkg";
import { RootArguments, Command, getLogger } from "../cli/command";

const debug = Debug("kosko:init");

function exists(path: string) {
  return new Promise(res => fs.exists(path, res));
}

export interface InitArguments extends RootArguments {
  force: boolean;
  path?: string;
}

export const initCmd: Command<InitArguments> = {
  command: "init [path]",
  describe: "Set up a new kosko directory",
  builder(argv) {
    return argv
      .option("force", {
        type: "boolean",
        describe: "Overwrite existing files",
        default: false,
        alias: "f"
      })
      .positional("path", { type: "string", describe: "Path to initialize" });
  },
  async handler(args) {
    const logger = getLogger(args);
    const path = args.path ? resolve(args.cwd, args.path) : args.cwd;
    logger.info("Initialize in", path);

    const exist = await exists(path);

    if (exist && !args.force) {
      throw new Error(
        "Already exists. Use --force to overwrite existing files."
      );
    }

    for (const name of ["components", "environments", "templates"]) {
      const dir = join(path, name);
      debug("Creating directory", dir);
      await makeDir(dir);
    }

    debug("Updating package.json");

    await writePkg(join(path, "package.json"), {
      dependencies: {
        "@kosko/env": "^0.1.0",
        "kubernetes-models": "^0.2.1"
      }
    });

    logger.success("Everything is set up");
  }
};
