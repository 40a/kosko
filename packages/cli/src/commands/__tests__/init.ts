import fs from "fs";
import { join } from "path";
import { Signale } from "signale";
import tempDir from "temp-dir";
import tmp from "tmp-promise";
import { promisify } from "util";
import { setLogger } from "../../cli/command";
import { initCmd, writeJSON } from "../init";

const readFile = promisify(fs.readFile);
const stat = promisify(fs.stat);

const logger = new Signale({ disabled: true });

beforeEach(() => jest.resetAllMocks());

describe("when the target exists", () => {
  let tmpDir: tmp.DirectoryResult;

  beforeEach(async () => {
    tmpDir = await tmp.dir({ dir: tempDir, unsafeCleanup: true });
  });

  afterEach(() => tmpDir.cleanup());

  test("should throw an error", async () => {
    const args = setLogger({ path: tmpDir.path } as any, logger);
    await expect(initCmd.handler(args)).rejects.toThrow("Already exists");
  });

  test("should proceed with --force flag", async () => {
    const args = setLogger({ path: tmpDir.path, force: true } as any, logger);
    await initCmd.handler(args);
  });

  describe("when package.json exists", () => {
    let pkgPath: string;

    beforeEach(async () => {
      pkgPath = join(tmpDir.path, "package.json");

      await writeJSON(pkgPath, {
        name: "foo",
        version: "1.2.3",
        dependencies: {
          debug: "3.2.1"
        }
      });
    });

    test("should update package.json", async () => {
      const args = setLogger({ path: tmpDir.path, force: true } as any, logger);
      await initCmd.handler(args);
      const content = await readFile(pkgPath, "utf8");
      expect(content).toMatchSnapshot();
    });
  });
});

describe("when path is not specified", () => {
  let tmpDir: tmp.DirectoryResult;

  beforeEach(async () => {
    tmpDir = await tmp.dir({ unsafeCleanup: true });
  });

  afterEach(() => tmpDir.cleanup());

  test("should use cwd instead", async () => {
    const args = setLogger({ cwd: tmpDir.path, force: true } as any, logger);
    await initCmd.handler(args);
  });
});

describe("success", () => {
  let tmpDir: tmp.DirectoryResult;
  let path: string;

  beforeAll(async () => {
    tmpDir = await tmp.dir({ unsafeCleanup: true });
    path = join(tmpDir.path, "target");

    const args = setLogger({ path } as any, logger);
    await initCmd.handler(args);
  });

  afterAll(() => tmpDir.cleanup());

  async function assertDir(...names: string[]) {
    const stats = await stat(join(path, ...names));
    expect(stats.isDirectory()).toBeTruthy();
  }

  async function assertFile(...names: string[]) {
    const content = await readFile(join(path, ...names), "utf8");
    expect(content).toMatchSnapshot();
  }

  test("should create components folder", () => assertDir("components"));

  test("should create environments folder", () => assertDir("environments"));

  test("should create templates folder", () => assertDir("templates"));

  test("should create package.json", () => assertFile("package.json"));

  test("should create kosko.toml", () => assertFile("kosko.toml"));
});
