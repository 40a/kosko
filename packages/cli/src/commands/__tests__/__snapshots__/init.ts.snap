// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`success should create package.json 1`] = `
"{
	\\"dependencies\\": {
		\\"@kosko/env\\": \\"^0.1.0\\",
		\\"kubernetes-models\\": \\"^0.2.1\\"
	}
}
"
`;

exports[`when the target exists when package.json exists should update package.json 1`] = `
"{
	\\"name\\": \\"foo\\",
	\\"version\\": \\"1.2.3\\",
	\\"dependencies\\": {
		\\"@kosko/env\\": \\"^0.1.0\\",
		\\"kubernetes-models\\": \\"^0.2.1\\"
	}
}
"
`;
